Testing: adding the ability to watch files/folders for TDD

samouri opened the pull request:
Implements #4922.

This pull request is to enable developers to pass along --watch (or -w) as a flag to the test-runner.   When watch mode is enabled, any change to js/jsx files will cause the test suite to be rerun.

Usage:
npm run test [folder] -- --watch

Test live: https://calypso.live/?branch=add/watch-tests


EDIT:

I should have added notes about implementation here so that people don't need to go clicking about:

File Watching with Mocha
--
Given that mocha allows for watching for files, my initial thought was to just pass through the watch flag to mocha like we do for a few of the other flags.

It turns out that Mocha does not support file-watching when used programmatically (as opposed to through their cli executable).  This is a bad thing in that we don't get file-watching for free but at the same time a good thing because they've set an example to follow for implementation.

- [watch utility function](https://github.com/mochajs/mocha/blob/8a37e016e5629b46709891740c74d5f0b69929e0/lib/utils.js#L168)
- [bussiness logic to rerun-mocha](https://github.com/mochajs/mocha/blob/b76989ceb8ca38b4ad9a2df4892581d96e8c4ba8/bin/_mocha#L343)

My current plan is to implement watching similarly to how mocha did (While keeping @aduth's logic for deciding which files to include).

aduth commented on test/runner.js:
```diff
@@ -18,37 +18,49 @@ const debug = require( 'debug' )( 'test-runner' ),
  * Internal dependencies
  */
 const boot = require( './boot-test' ),
+	utils = require( './utils.js' ),
```
Can omit file extension `.js`

samouri replied:
will do!

aduth commented on test/runner.js:
```diff
@@ -18,37 +18,49 @@ const debug = require( 'debug' )( 'test-runner' ),
-	ui: 'bdd',
-	reporter: program.reporter
-} );
+let getMocha = function() {
```
We encourage `const` over `let` unless you're reassigning the value ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#variable-declarations))

samouri replied:
Thank you for including references! I'll modify the ones that make sense.

aduth commented on test/runner.js:
```diff
@@ -18,37 +18,49 @@ const debug = require( 'debug' )( 'test-runner' ),
 
-mocha.suite.beforeAll( boot.before );
-mocha.suite.afterAll( boot.after );
+	let testSuiteLoaderPath = path.join( __dirname, 'load-suite.js' );
```
`const` over `let`

samouri replied:
will do

aduth commented on test/runner.js:
```diff
@@ -87,10 +99,42 @@ if ( program.nodeTotal > 1 ) {
+
+	let rerun = () => {
+		console.log( chalk.green( 'Running tests at: ' ),
+									chalk.yellow( new Date( ) ) );
```
- Some wild tabbing going on here, at most we'd want indented one more than the previous line, else consider including in the same line, or having arguments indented one-per-line.
- Space in `new Date` constructor unnecessary

samouri replied:
1) Thanks for noticing that.  I'll  fix it up
2) I'm pretty sure I was getting eslint warnings without the space.  I'll double check

aduth replied:
>I'm pretty sure I was getting eslint warnings without the space. I'll double check

Spaces are necessary when arguments are present, but can be left out otherwise.

- [Guideline](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls)
- [ESLint rule](http://eslint.org/docs/rules/space-in-parens) (we configure to "always")

aduth commented on test/runner.js:
```diff
@@ -87,10 +99,42 @@ if ( program.nodeTotal > 1 ) {
+									chalk.yellow( new Date( ) ) );
+		runAgain = false;
+
+		runner = getMocha().run( function( ) {
```
Space in empty function arguments unnecessary

samouri replied:
Got it

aduth commented on test/runner.js:
```diff
@@ -87,10 +99,42 @@ if ( program.nodeTotal > 1 ) {
+	// start the first test-run independently of watch
+	rerun();
+
+	let watchFiles = utils.files( process.cwd( ), [ 'js', 'jsx' ] );
```
- `const` over `let`
- Space in `process.cwd` call unnecessary

samouri replied:
Will do

aduth commented on test/utils.js:
```diff
@@ -0,0 +1,70 @@
+import { watchFile, readdirSync, statSync } from 'fs';
```
Some inconsistent usage of ES6 in this file.

- `module.exports` should be replaced with [`export`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export), e.g. `export function watch( files, fn ) {`
- A few `var`, to be replaced with `const` or `let` ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#variable-declarations))

samouri replied:
I didn't do as good a job as I should have when taking Mocha's code and converting styles

aduth commented on test/utils.js:
```diff
@@ -0,0 +1,70 @@
+ * @return {boolean} true if should ignore
+ */
+function ignored( fp ) {
+	return ! ~ignore.indexOf( fp );
```
I'd consider this ["clever" code](http://programmers.stackexchange.com/questions/25276/why-is-cleverness-considered-harmful-in-programming-by-some-people/25281) and would prefer to see written as `-1 !== ignore.indexOf( fp )`. In many cases we go as far as to pull in [`_.includes`](https://lodash.com/docs#includes) which is nicely obvious to its intent.

aduth replied:
Ah, now I'm seeing that these are pulled from Mocha. Can we not just import directly from the package?

samouri replied:
1. I fully agree and will switch to _.includes
2. I'm pretty curious about this.  Technically we could directly import from Mocha but that feels brittle to me.  We become tied to Mocha's implementation (undocumented functionality that they have no promise to support/keep).  Let me know what you think!

aduth replied:
>We become tied to Mocha's implementation (undocumented functionality that they have no promise to support/keep). Let me know what you think!

This is a fair point, and maybe we should consider exploring other utilities that offer this functionality as their stated purpose (e.g. [gaze](npmjs.com/package/gaze)). That said, bringing in a new dependency has its own trade-offs in terms of maintainability. Since there's only a handful of functions here, it might be reasonable to copy them from Mocha, though as stated above, there's several styling concerns to address in doing so.

aduth commented on test/utils.js:
```diff
@@ -0,0 +1,70 @@
+ * @return {Array} of files
+ */
+exports.files = function( dir, ext, ret ) {
+	ret = ret || [ ];
```
Space in empty array unnecessary

aduth replied:
For clarity's sake, we tend to discourage abbreviated variable names in favor of their more verbose counterpart ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#naming-conventions)).

samouri replied:
I'll fix up the names (and potentially exra spaces depending on eslint)

aduth commented on test/runner.js:
```diff
@@ -87,10 +99,42 @@ if ( program.nodeTotal > 1 ) {
+	utils.watch( watchFiles, () => {
+		runAgain = true;
+		if ( runner ) {
+			runner.abort();
```
What happens to reach this point? Might be worth adding a comment to clarify.

samouri replied:
I'll see if I can either simplify the logic and if not add a comment

aduth commented:
Thanks for tackling this, will be a nice addition for testing!

gziolo commented:
First of all, thank you very much for taking care of it :)

I did several tests locally. It looks like we should purge all cached files to make it work. This is what happens when I execute `npm run test-client client/state/ -- -w`:

![watch-tests](https://cloud.githubusercontent.com/assets/699132/16050865/91711c64-325d-11e6-8030-6d2ffa93aaac.gif)

samouri commented:
@gziolo , What you found is interesting.  It actually happens even without modifying the file so I don't think it is related to purging the cache. 

1. If you run `npm run test-client client/state/analytics -- -w` there is no issue and it behaves as you would expect.
2. If you run `npm run test-client client/state -- -w` then it fails on the successive runs no matter what.

I'll figure out whats going on

samouri commented:
Notes on latest PR:
1. I've removed the utils file entirely and am using glob + direct calls to watchFile
2. I've investigated the difference between fs.watchFile and fs.watch.  While fs.watch is currently recommended by Node and is more performant, it still has compatibility issues and I noticed it consistently borking on my box (would not pick up file modifications).  Given all that I stuck with fs.watchFile
3. Simplified all of the logic regarding rerunning.  I think I've found a simpler approach than what mocha does --
3.1 Mocha creates extra stateful variable about runAgain (as I had done in the first commit), in order to handle non-instant aborts.  I thought it was cleaner to add an explicit wait after an abort.
4. Fixed the issue @gziolo noticed. I think it had something to do with not invalidating node_modules -- specifically nock.

gziolo commented on test/runner.js:
```diff
@@ -12,46 +12,60 @@ const debug = require( 'debug' )( 'test-runner' ),
 
-files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
-files = files.reduce( ( memo, filePath ) => {
+testFiles = program.args.length ? program.args : [ process.env.TEST_ROOT ];
```
We rather should move all logic related to `testFiles` inside `getMocha` to make sure Mocha is able to pick up new files and skip processing removed files.

gziolo replied:
I can confirm that it fails when I remove test file.

samouri replied:
advanced feature -- I like it
+1

gziolo commented on test/runner.js:
```diff
@@ -80,17 +94,42 @@ files = files.reduce( ( memo, filePath ) => {
-	process.on( 'exit', function() {
-		process.exit( failures ); //eslint-disable-line no-process-exit
+	runner = getMocha().run( () => {
+		runner = null;
```
I think we still need `process.on( 'exit', ... )` to give feedback to our Continuos Integration (Circle CI) that tests have failed.

samouri replied:
gotcha -- I'll add it back in. I didn't know what it was for so thanks for catching that before I broke CI

gziolo commented on test/runner.js:
```diff
@@ -12,46 +12,60 @@ const debug = require( 'debug' )( 'test-runner' ),
+	 * We invalidate the cache because Mocha has trouble running a test-suite
+	 * more than once:  https://github.com/mochajs/mocha/issues/995
+	 */
+	Object.keys( require.cache ).forEach( ( key ) => {
```
Wouldn't it better to call it only when Mocha execution was triggered by watch feature? We don't need that for the first Mocha run.

samouri replied:
You  mean that purging cache is not necessary on the first run?  While I agree of course, purging the cache each time makes the code more simple.  I ran it both ways with the time command to see if there was a noticeable speed difference.  The random difference in successive runs was more significant than the difference caused by cache-purging.   In order to get a more accurate picture for if there is a statistically significant performance hit I would need to grab the avg over many runs (seems like overkill / over-optimization). 

Given the research I performed, I would prefer to run the cache-purge every run for simplicity.
 

Here were some sample times on my 2013 mac-air:

Without purging cache:

<pre> 
real	0m5.166s
user	0m4.712s
sys	    0m0.554s
</pre>

With purging cache:
<pre> 
real	0m5.187s
user	0m4.755s
sys	    0m0.520s
</pre>

gziolo replied:
In the linked Mocha examples they purge cache only for reruns. I must disagree that it makes our code more simple, because you can move all logic to a different place in a file (line 138 at the moment) and it will work exactly the same. My concern here is that purging the cache isn't the part of usual Mocha setup, but current implementation can lead to that conclusion for someone who doesn't know the context.

samouri replied:
Sure, I'll move it there

gziolo commented:
I can confirm, that Mocha reporting no longer works properly. In this branch (https://circleci.com/gh/Automattic/wp-calypso/17949) we have:

![screen shot 2016-06-15 at 08 46 16](https://cloud.githubusercontent.com/assets/699132/16070433/a24959e8-32d5-11e6-99be-81b87ce1a2e1.png)

while on master (https://circleci.com/gh/Automattic/wp-calypso/17964) we have:

![screen shot 2016-06-15 at 08 47 00](https://cloud.githubusercontent.com/assets/699132/16070448/c0700516-32d5-11e6-8bf4-1bca95f00ae5.png)

gziolo commented on test/runner.js:
```diff
@@ -80,17 +94,42 @@ files = files.reduce( ( memo, filePath ) => {
+if ( program.watch ) {
+	console.log( chalk.green( 'Watch mode enabled' ) );
+
+	const watchFiles = glob.sync( '**/*.@(js|jsx)', { ignore: 'node_modules/**' } );
```
I see now that it won't cover the scenario when I create new test file. Do they even care about it in Mocha default test runner? I'm not saying we must have it, just curious :D

samouri replied:
I don't think Mocha supports it: https://github.com/mochajs/mocha/blob/b76989ceb8ca38b4ad9a2df4892581d96e8c4ba8/bin/_mocha#L343-L406

aduth replied:
Downside of `sync` here is that it blocks the initial `runMocha` below.

samouri replied:
That's a good point. I'll unsync it

gziolo commented:
I left some comments. We still have some issues to fix, but it looks like we are close to having watch feature implemented, yay 🎉  Really awesome improvement ;)

samouri commented:
Notes on latest PR:
--
1.  setup.js and load-suite.js were not doing anything anymore so I removed them
2. removed run-mocha. It would be better to call runner.js directly in webstorm rather than go through the Mocha executable 
3. I fixed CI integration (added back process.exit and fixed a typo in reporter)
4. Added support for adding/remove testfiles while watching (although added files will not re-trigger a rerun of the test-suite even though they will get tested)

gziolo commented on test/run-mocha.js:
```diff
@@ -1,10 +0,0 @@
-/**
```
We should check with @umurkontaci if we can remove that file. I guess we need to keep it, because it hooks `before` and `after` function calls which contain logic essential for some tests to work properly.

umurkontaci replied:
The purpose of this file was to enable Mocha integration with IDEs. We were able to run Mocha tests directly inside WebStorm with it.

Later I noticed that some of the logic inside `load-suite` has changed and broke this. I'd love see this working again but for the time being I think it's safe to remove this file since it's not working.

gziolo replied:
Thanks for update, I'll create new issue to fix that :)

gziolo replied:
Done: #6076 :)

gziolo commented on test/runner.js:
```diff
@@ -10,87 +9,138 @@ require( 'babel-register' );
+			let abortedCheck = setInterval( () => {
+				if ( runner == null ) {
+					clearInterval( abortedCheck );
+					runMocha();
```
This is where you can purge cache.

samouri replied:
moved

gziolo commented on test/runner.js:
```diff
@@ -10,87 +9,138 @@ require( 'babel-register' );
-	files = files.filter( ( file, index ) => {
-		return index % program.nodeTotal === program.nodeIndex;
+		// Inform CI of failures
+		process.on( 'exit', () => {
```
We should move it to the end of the script. Otherwise we add another even listener each time we rerun tests.

samouri replied:
I spoke with @blowery, this is better now 👍

gziolo commented:
@aduth @blowery @Tug: it looks like we are close to having watch feature incorporated. Do you have any more feedback?

gziolo commented on test/load-suite.js:
```diff
@@ -1,17 +0,0 @@
-		if ( folderName === 'test' && Array.isArray( folderConfig ) ) {
-			folderConfig.forEach( fileName => require( `${path}test/${fileName}` ) );
-		} else {
-			describe( folderName, () => {
```
@aduth: I remember that I followed your with how you approached Redux state tests to remove `Makefile` files there. 
Are there any reason to keep tree like structure? I think we can remove it now to simplify our code.

aduth replied:
>Are there any reason to keep tree like structure?

I most like it in that it helps clarify how parts of the application relate to each other by their folder hierarchy, but I don't think it's absolutely necessary.

aduth commented on test/runner.js:
```diff
@@ -10,87 +9,133 @@ require( 'babel-register' );
-} );
+let getTestFiles = () => {
+	let testFiles = program.args;
+	if ( isEmpty( program.args ) ) {
```
More food for thought since it doesn't bother me much, but seems `! program.args.length` would work just as well here. `isEmpty` is nice in that it's human-readable, but knowing that it supports both objects and arrays makes me a bit more uncertain what it is you're testing.

samouri replied:
isn't that a problem with `isEmpty` as a whole and not its specific usage here?

aduth commented on test/runner.js:
```diff
@@ -10,87 +9,133 @@ require( 'babel-register' );
+			}
+			// aborting is not instant, so wait for it to finish
+			// then call run the test-suite again
+			let abortedCheck = setInterval( () => {
```
Lint warnings after rebase:

```
./node_modules/.bin/eslint test/runner.js 

/Users/andrew/Documents/Code/wp-calypso/test/runner.js
   35:5  warning  `getTestFiles` is never modified, use `const` instead  prefer-const
  125:8  warning  `abortedCheck` is never modified, use `const` instead  prefer-const

✖ 2 problems (0 errors, 2 warnings)
```

Related: #6093

samouri replied:
I figured this code was isolated enough that I wouldn't need to rebase very often but I guess I was wrong :)

aduth commented on test/runner.js:
```diff
@@ -10,87 +9,133 @@ require( 'babel-register' );
+			// aborting is not instant, so wait for it to finish
+			// then call run the test-suite again
+			let abortedCheck = setInterval( () => {
+				if ( runner == null ) {
```
Hmm, wonder why we have an exception here for triple equals:

https://github.com/Automattic/calypso-pre-oss/blob/master/docs/coding-guidelines/javascript.md#equality

Seems `! runner` could cover the broader use-case. Granted, also catches `0` or `false`, but I'm curious what use case we'd have for checking against both `null` and `undefined`. In other words, no reason this couldn't be `===` here.

samouri replied:
I'll change to ===

samouri replied:
I'll change it to `===`

aduth commented on test/runner.js:
```diff
@@ -10,87 +9,133 @@ require( 'babel-register' );
+					} );
+					runMocha();
+				}
+			}, 25 );
```
How'd you come to `25` ?

samouri replied:
Its completely arbitrary at this point.  Let me know if you have any ideas on if/how I should pick a number.

aduth commented:
Hrm, encountered a bug I think.

__Steps to reproduce:__

1. `npm run test-client client/state/posts -- --watch`
2. Note all tests pass
3. Save a file
4. Tests are rerun, but this time 14 failures

They all seem to be similar to this one:

```
1) actions #requestPostCounts() should dispatch request action when thunk triggered:
     TypeError: (0 , _chai.expect)(...).to.have.been.calledWith is not a function
      at Context.<anonymous> (actions.js:71:31)
```

samouri commented:
@aduth, you are right.  When I moved `const boot = require( './boot-test' ); ` outside of `getMocha()` I broke it 👎 .  Action items for next commit:

1. Move the require back to inside of getMocha
2. Fix the new eslint warnings
3. unsyncify the glob.sync call inside of the program.watch branch

blowery commented on test/runner.js:
```diff
@@ -10,87 +9,134 @@ require( 'babel-register' );
+};
+
+if ( program.watch ) {
+	console.log( chalk.green( 'Watch mode enabled' ) );
```
we should really use `https://www.npmjs.com/package/chokidar` for this. `fs.watch` has a bunch of warts.

blowery replied:
plus it'll catch adds.

samouri replied:
@blowery, 
Using chokidar would make the code more performant CPU-wise, and allow for catching adds.
To the extent that I am using `watchFile` those are the only benefits I know of.  
Personally, I don't think that is worth pulling in a dependency but if you feel otherwise I'll change it.

samouri commented:
@everyone on this thread (@aduth, @blowery, @gziolo, @umurkontaci)

After working on it for a bit, and especially after seeing how much of a hack it is to make Mocha play nicely with file-watching, I don't think we should be implementing watch as a flag to the test-runner.

All of the benefit of test-watching can be gained by using something like [chokidar-cli](https://www.npmjs.com/package/chokidar-cli) and a simple command like: `chokidar 'client/**/*.js' test/**/*.js -c 'pkill -f runner.js; npm run test-client'` (this already works today and was pointed out to me by @blowery).  We don't need to add any calypso code.  

There is a healthy discussion about the problems with Mocha watching here: https://github.com/mochajs/mocha/issues/139.

As for the current implementation, it's more half-baked than I realized.  There is currently a memory leak and it eventually crashes if watching for long enough. I'm not sure exactly what is going on there yet

Let me know what you guys think about the two alternatives:
1. Add documentation + potentially an extra npm target for using `chokidar-cli` (my preference)
2. Continue to debug/implement the `--watch` flag for runner.js

aduth commented:
I like the idea of an additional NPM script for test watching, even if it's just to pass options to the Chokidar CLI. If we pull it in as a devDependency, we also won't need to require it be installed globally by the developer.

blowery commented:
> I like the idea of an additional NPM script for test watching, even if it's just to pass options to the Chokidar CLI.

Agreed. Maybe have the script prepopulate watching the entire client / server area for changes (or just the whole project?) and pass it the folder you want to watch, or a glob. So usage would be something like  `npm run test-watch -- client/state/reader`

samouri commented:
@blowery, I found it difficult to make passing in a folder very clean because of the nature of how arg-passing in NPM works.  Let me know if you know of a better way: #6321

samouri commented:
Closing this branch because we decided not to go forward with it as opposed to a simpler solution like the one proposed in #6321.
